common/src/graph_collision.rs

```diff
@@ -12,39 +16,121 @@ use crate::{
+    // the corresponding neighboring chunk will also be used for collision checking.
+    let klein_lower_boundary = collider_radius.tanh();
+    let klein_upper_boundary =
+        ((Vertex::chunk_to_dual_factor() as f32).atanh() - collider_radius).tanh();
+
+    // Breadth-first search loop
+    while let Some((chunk, node_transform)) = chunk_queue.pop_front() {
+        let Chunk::Populated {
+                voxels: ref voxel_data,
+                ..
+            } = graph[chunk] else {
+                // Collision checking on unpopulated chunk
+                return Err(SphereCastError::OutOfBounds);
+            };
+        let local_ray = chunk.vertex.node_to_dual().cast::<f32>() * node_transform * ray;
+
+        // Check collision within a single chunk
+        let current_tanh_distance = hit.as_ref().map_or(tanh_distance, |hit| hit.tanh_distance);
+        hit = chunk_sphere_cast(
+            collider_radius,
+            voxel_data,
+            layout,
+            &local_ray,
+            current_tanh_distance,
+        )
+        .map_or(hit, |hit| {
+            Some(GraphCastHit {
+                tanh_distance: hit.tanh_distance,
+                chunk,
+                normal: math::mtranspose(&node_transform)
+                    * chunk.vertex.dual_to_node().cast()
+                    * hit.normal,
+            })
+        });
+
+        // Compute the Klein-Beltrami coordinates of the ray segment's endpoints. To check whether neighboring chunks
+        // are needed, we need to check whether the endpoints of the line segments lie outside the boundaries of the square
+        // bounded by `klein_lower_boundary` and `klein_upper_boundary`.
+        let klein_ray_start = na::Point3::from_homogeneous(local_ray.position).unwrap();
+        let klein_ray_end =
+            na::Point3::from_homogeneous(local_ray.ray_point(current_tanh_distance)).unwrap();
+
+        // Add neighboring chunks as necessary, using one coordinate at a time.
+        for axis in 0..3 {
+            // Check for neighboring nodes
+            if klein_ray_start[axis] <= klein_lower_boundary
+                || klein_ray_end[axis] <= klein_lower_boundary
+            {
+                let side = chunk.vertex.canonical_sides()[axis];
+                let next_node_transform = side.reflection().cast::<f32>() * node_transform;
+                // Crude check to ensure that the neighboring chunk's node can be in the path of the ray. For simplicity, this
+                // check treats each node as a sphere and assumes the ray is pointed directly towards its center. The check is
+                // needed because chunk generation uses this approximation, and this check is not guaranteed to pass near corners.
+                let ray_node_distance = (next_node_transform * ray.position).w.acosh();
+                let ray_length = current_tanh_distance.atanh();
+                if ray_node_distance - ray_length
```

Comment:
Should we also account for the collider radius here?

Reply:
That is handled by `> dodeca::BOUNDING_SPHERE_RADIUS as f32 + collider_radius`. However, I should take another look at which terms are on which side of the inequality sign for minimal confusion.

Reply:
Whoops, not sure how I missed that. LGTM.

Reply:
There isn't a clear reason why `collider_radius` is on the right side instead of the left, so it is pretty well hidden because of that. I think it's a holdover from when I visualized this as ray tracing with inflated voxels instead of sphere casting with regular voxels.

I've moved it over to the left side where it's easier to see.